concurrency/BaseFilterAdapter.java

```diff
@@ -0,0 +1,645 @@
+
+        int count = getPartitionCount();
+        int limit = 0;
+
+        // Since we will be changing several partitions at once, hold the data change
+        // notifications
+        setNotificationsEnabled(false);
+        try {
+            // The filter has loaded results for the default partition too.
+            if (defaultPartitionCursor != null && getPartitionCount() > 0) {
+                changeCursor(0, defaultPartitionCursor);
+            }
+
+            int defaultPartitionCount = (defaultPartitionCursor == null ? 0
+                : defaultPartitionCursor.getCount());
+
+            limit = mPreferredMaxResultCount - defaultPartitionCount;
+
+            // Show non-default directories as "loading"
+            // Note: skipping the default partition (index 0), which has already been loaded
+            for (int i = 1; i < count; i++) {
+                DirectoryPartition partition = (DirectoryPartition) getPartition(i);
+                partition.constraint = constraint;
+
+                if (limit > 0) {
+                    if (!partition.loading) {
+                        partition.loading = true;
+                        changeCursor(i, null);
+                    }
+                } else {
+                    partition.loading = false;
+                    changeCursor(i, null);
+                }
+            }
+        } finally {
+            setNotificationsEnabled(true);
+        }
+
+        // Start search in other directories
+        // Note: skipping the default partition (index 0), which has already been loaded
+        for (int i = 1; i < count; i++) {
+            DirectoryPartition partition = (DirectoryPartition) getPartition(i);
+            if (partition.loading) {
+                mHandler.removeMessages(MESSAGE_SEARCH_PENDING, partition);
+                Message msg = mHandler.obtainMessage(MESSAGE_SEARCH_PENDING, i, 0, partition);
+                mHandler.sendMessageDelayed(msg, MESSAGE_SEARCH_PENDING_DELAY);
+                partition.filter = new DirectoryPartitionFilter(i, partition.directoryId);
+                partition.filter.setLimit(limit);
+                // Add the query to queue. IF the queue already has an entry for this directory,
+                // the constraint is updated. No need to add it again.
+                if (!mPendingDirectoryQueryQueue.contains(partition)) {
+                    mPendingDirectoryQueryQueue.add(partition);
+                }
+                //Kick off the processing of the queue
+                processGalQueryQueue();
+            }
+        }
+    }
+
+    // Helper method to process the Gal queries. First checked if there are any available permits,
+    // and schedules a query only if there are.
+    private void processGalQueryQueue() {
+        while (!mPendingDirectoryQueryQueue.isEmpty()) {
+            if (mParallelQueryPermits.tryAcquire()) {
+                DirectoryPartition partition = mPendingDirectoryQueryQueue.remove();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentLinkedQueue`, but your usage of `isEmpty()` and `remove()` may not be thread-safe at lines: **529 and 531**. If the `ConcurrentLinkedQueue` contains only **one element** and two threads perform this **same check at the same time** , the second thread&#39;s call to `remove()` will throw `NoSuchElementException`.

**Fix**  
 Consider using `poll()` (does not throw an exception, but it may return `null`, which you can check for) or handling the exception.